Design/WorkspacePropertyWithIndex.md

```diff
@@ -0,0 +1,130 @@
+  - Some algorithms do not validate for duplicate spectrum numbers and double-process the corresponding spectra (see [16651](https://github.com/mantidproject/mantid/issues/16651)).
+  - Validations are not unit-tested to a good extent.
+
+## Design Goals
+ - Provide a property type which can specify a workspace and allowed (and predefined) input index types.
+ - Unified behaviour on error (when bad indices are provided).
+ - Hide index translations from the developer (this will take place internally within the property).
+ - Automatically generate the required GUI input controls based on the specified index types.
+ - Return the workspace and correct index set simultaneously in one call to `PropertyManager::getProperty`. 
+ - Consistent and unit-tested translation without code duplication.
+ - Consistent and unit-tested validation without code duplication.
+ - Consistent and simple way of defining index properties without code duplication.
+ - Consistent user interface.
+ 
+## Implementation
+
+### Basics
+
+- Extend the `WorkspaceProperty` type.
+- Own an `ArrayProperty` which can make use of `ListPropertyWidget` for capturing list data from the GUI.
+- Add a set of flags which control allowed index types (for use in the constructor).
+- Use a `PropertyWithValue<std::string>` to capture the selected index type.
+
+### Details
+
+1. The `WorkspacePropertyWithIndex` type constructor will accept the following flags (or combinations of flags):
+	- `IndexType::SpectrumNumber` for list of global<sup>1</sup> spectrum numbers typically from 1-N.
+	- `IndexType::WorkspaceIndex` for list of global workspace indices typically from 0-N.
+	- `IndexType::DetectorID` for list of detector ids within the instrument.
+2. Users will be restricted to entering only one of the allowed index types at the GUI level. 
+3. Validation will occur in `IndexInfo::makeIndexSet`(which throws for bad indices), not in every algorithm.
+4. The `ArrayProperty` will be exposed e.g `WorkspacePropertyWithIndex::getArrayProperty` in order to make use of `ListPropertyWidget`.
+5. Overload function call operator to return `std::pair`.
+6. `IndexInfo` will be used to translate the indices at the point of access i.e. `WorkspacePropertyWithIndex::operator()()`.
+7. There must be a few modifications to the GUI generation code in order to support this type:
+	-  `PropertyWidgetFactory` may need to handle `WorkspacePropertyWithIndex` specifically.
+	-  May need to introduce a concept of a group of property widgets.
+
+<sup>1</sup><sub>The distinction is made here between local (a subset of all data partitioned on an MPI Rank) and global (entire data set). This will not be visible at the GUI or client API level.</sub>
+
+### Example usage
+
+**Property Declaration**:
+
+Declaring the property in C++:
+```cpp
+//Property name may be fixed?
+declareProperty(make_unique<WorkspacePropertyWithIndex<MatrixWorkspace>>("InputWorkspaceWithIndex", IndexType::SpectrumNumber|IndexType::WorkspaceIndex));
```

Comment:
I think as @SimonHeybrock says may we can sort this out by having `MatrixWorkspace` as a default template parameter and default `IndexType::WorkspaceIndex`. That would cover most cases and then only some places would need the explicit version.

`PropertyManager` currently managers pointers so `make_unique` just ensures the pointer has been created - I'm not sure how best to deal with that. It would be nice just to have the `PropertyManager` move construct a new object and handle that.